upb/mini_table.c

```diff
@@ -0,0 +1,937 @@
+    encoded_type += kUpb_EncodedType_RepeatedBase;
+  }
+  ptr = upb_MtDataEncoder_Put(e, ptr, encoded_type);
+  if (!ptr) return NULL;
+
+  uint32_t encoded_modifiers = 0;
+  if (modifiers & kUpb_FieldModifier_IsProto3Singular) {
+    encoded_modifiers |= kUpb_EncodedFieldModifier_IsProto3Singular;
+  }
+  if (modifiers & kUpb_FieldModifier_IsRequired) {
+    encoded_modifiers |= kUpb_EncodedFieldModifier_IsRequired;
+  }
+  if ((modifiers & kUpb_FieldModifier_IsPacked) !=
+      (in->msg_mod & kUpb_MessageModifier_DefaultIsPacked)) {
+    encoded_modifiers |= kUpb_EncodedFieldModifier_IsUnpacked;
+  }
+  if (encoded_modifiers) {
+    ptr = upb_MtDataEncoder_PutBase92Varint(e, ptr, encoded_modifiers,
+                                            kUpb_EncodedValue_MinModifier,
+                                            kUpb_EncodedValue_MaxModifier);
+  }
+  return ptr;
+}
+
+char* upb_MtDataEncoder_StartOneof(upb_MtDataEncoder* e, char* ptr) {
+  upb_MtDataEncoderInternal* in = upb_MtDataEncoder_GetInternal(e, ptr);
+  if (in->oneof_state == kUpb_OneofState_NotStarted) {
+    ptr = upb_MtDataEncoder_Put(e, ptr, upb_FromBase92(kUpb_EncodedValue_End));
+  } else {
+    ptr = upb_MtDataEncoder_Put(
+        e, ptr, upb_FromBase92(kUpb_EncodedValue_OneofSeparator));
+  }
+  in->oneof_state = kUpb_OneofState_StartedOneof;
+  return ptr;
+}
+
+char* upb_MtDataEncoder_PutOneofField(upb_MtDataEncoder* e, char* ptr,
+                                      uint32_t field_num) {
+  upb_MtDataEncoderInternal* in = upb_MtDataEncoder_GetInternal(e, ptr);
+  if (in->oneof_state == kUpb_OneofState_EmittedOneofField) {
+    ptr = upb_MtDataEncoder_Put(
+        e, ptr, upb_FromBase92(kUpb_EncodedValue_FieldSeparator));
+    if (!ptr) return NULL;
+  }
+  ptr = upb_MtDataEncoder_PutBase92Varint(e, ptr, field_num, upb_ToBase92(0),
+                                          upb_ToBase92(63));
+  in->oneof_state = kUpb_OneofState_EmittedOneofField;
+  return ptr;
+}
+
+const upb_MiniTable_Field* upb_MiniTable_FindFieldByNumber(
+    const upb_MiniTable* table, uint32_t number) {
+  int n = table->field_count;
+  for (int i = 0; i < n; i++) {
+    if (table->fields[i].number == number) {
+      return &table->fields[i];
+    }
+  }
+  return NULL;
+}
+
+/** Data decoder **************************************************************/
+
+typedef enum {
+  kUpb_LayoutItemType_Field,       // Non-oneof field data.
+  kUpb_LayoutItemType_OneofField,  // Oneof field data.
+  kUpb_LayoutItemType_OneofCase,   // Oneof case.
+} upb_LayoutItemType;
+
+#define kUpb_LayoutItem_IndexSentinel ((uint16_t)-1)
+
+typedef struct {
+  // Index of the corresponding field.  When this is a oneof field, the field's
+  // offset will be the index of the next field in a linked list.
+  uint16_t field_index;
+  upb_FieldRep rep;
+  upb_LayoutItemType type;
+
+  // Used for temporary storage while assigning offsets (internal only).
+  uint16_t offset;
```

Comment:
I suspect you are wasting some space in padding.  Maybe put this up with the index?

Reply:
Done (and removed obsolete comment).